src/ClassificationTree.jl

```diff
@@ -0,0 +1,200 @@
+"""
+function build_tree(data::Matrix{T}, labels::Vector{L}, max_depth::Int, min_samples_split::Int, depth::Int=0) where {T, L}
+    
+    # If max_depth is reached or if the data can not be split further, return a leaf
+    if length(labels) < min_samples_split || (max_depth != -1 && depth >= max_depth) 
+        return Leaf{L}(labels)                                                       
+    end
+    
+    # Get the best split from the respective split_criterion
+    feature_index, split_value = find_best_split(data, labels)
+    # Random values for testing purposes
+    # feature_index = rand((1:size(data, 2)))
+    # split_value = data[rand((1:size(data, 1))), feature_index]
+
+    # Create the mask on the data
+    if isa(split_value, Number)
+        left_mask = data[:, feature_index] .< split_value
+        right_mask = data[:, feature_index] .>= split_value
+    else
+        left_mask = data[:, feature_index] .!= split_value
+        right_mask = data[:, feature_index] .== split_value
+    end
+
+    # If the data can not be split further, return a leaf
+    if allequal(left_mask) || allequal(right_mask)
+        return Leaf{L}(labels)
+    end
+
+    # Compute the data and labels for the child nodes
+    left_data, right_data = data[left_mask, :], data[right_mask, :]
+    left_labels, right_labels = labels[left_mask], labels[right_mask]
+
+    # Create the node with the computed attributes
+    node = Node(data, labels)
+    node.feature_index = feature_index
+    node.split_value = split_value
+
+    # Build the subtrees for the node
+    node.left = build_tree(left_data, left_labels, max_depth, min_samples_split, depth + 1)
+    node.right = build_tree(right_data, right_labels, max_depth, min_samples_split, depth + 1)
+
+    # Return the node
+    return node
+end
+
+"""
+    fit(tree::ClassificationTree)
+
+Compute the tree structure.
+"""
+function fit(tree::ClassificationTree)
+    tree.root = build_tree(tree.data, tree.labels, tree.max_depth, tree.min_samples_split)
+end
+
+"""
+    predict(tree::ClassificationTree, data::Matrix{T})
+
+Returns the prediction of the ClassificationTree for a list of datapoints.
+
+`data` contains the datapoints to predict.
+"""
+function predict(tree::ClassificationTree, data::Matrix{T}) where {T}    
+    predictions = []
+    
+    for i in 1:size(data, 1)
+        node = tree.root
+        while !isa(node, Leaf)
+            if isa(node.split_value, Number)
+                if data[i, node.feature_index] < node.split_value
+                    node = node.left
+                else
+                    node = node.right
+                end
+            else
+                if data[i, node.feature_index] != node.split_value
+                    node = node.left
+                else
+                    node = node.right
+                end
+            end
+        end
+        # Get the label that occurs the most and add it to predictions
+        push!(predictions, mode(node.values))
```

Comment:
Maybe use the same Helper function within predict like below

for i in 1:size(data, 1)
    push!(predictions, traverse_tree(tree.root, data[i, :]))
end